015-kafka-topic-encryption.md

```diff
@@ -0,0 +1,169 @@
+If a topic matches a topic in the encryption policy, the contents of the _value_ field of each record is encrypted and replaced with the corresponding ciphertext. Metadata about the encryption (e.g., encrypted key or key reference, nonce) is added to the record's headers in order that this information is available later for decryption.
+
+The encryption metadata is currently stored in Kafka record headers. Record headers are optional name value/pairs stored in an array associated with the record. Record header names are not required to be unique, thus name "collisions" are possible. Ordering and versioning are used to avoid problems arising from names used for metadata which coincidentally equal names chosen by a Kafka client. Ordering means that the record header is rewritten such that the encryption-related headers
+occur first in the array. The first header always contains a version from which the exact
+metadata field names can be derived. The decryption algorithm therefore can discern between
+headers added by the encryption module and those added by the client. The version field
+is enables future migration and backwards compatibility should changes in semantics occur such as modified message formats or encryption options.
+
+Although not part of our reference implementation, the encryption of header values is potentially an additional requirement to consider. In this case, policy would be extended to indicate that specific or all client header values be encrypted. Since record headers are already processed during encryption, and since the algorithm can discriminate between encryption and client headers, it should be just an incremental extension encrypt header data as well. Encrypting header values will not impinge on storing encryption metadata in headers as only client headers will be encrypted. 
+Encrypting header names as well would likewise be a relatively simple extension however we recommend further investigating the full implications of encrypting header names.
+
+### Embedding Topic Encryption in a Proxy
+A proxy can be deployed as a free-standing process or as a sidecar in a Kubernetes pod. 
+In both cases, the proxy needs to intercept and de-serialize Kafka messages, encrypt/decrypt the relevant records, 
+and reserialize them into a new message. The overhead of a proxy potentially has an effect on performance.
+It also means that the proxy version must always be in phase with that of the broker.
+A proxy-based solution however has the advantage that both Kafka client and broker are unaware
+of the proxy and do not require any modification or configuration to support encryption at rest.
+
+Whether the proxy runs under the control of the client or the broker is considered a configuration
+issue, i.e. exactly the same proxy would be used in both cases, but when under the control of the
+client it would be their responsibility to ensure that producers/consumers are passing through a proxy with the same configuration.
```

Comment:
Client identity might be a problem here. When the proxy is just proxying a single client it can forward the `SaslHandshake` and `SaslAuthenticate` messages, so SASL authentication (configure on the client) should just work. Likewise TLS authentication would work fine, because the proxy can be configured with the TLS keystore and truststore which the client would have used. But it gets more complicated when more than a single client it using the same proxy (and that could even happen in the sidecar deployment, if there were multiple client instances being used in the main container of the pod). You probably don't want to worry about this now, which is fine, but I think we should be explicit about that.

Similarly there can be issues with client quotas if the proxy is serving multiple clients but using a single identity for its connection to the broker.

Reply:
You raise important integration issues which will have to be addressed. Indeed, they may vary depending on how the proxy is deployed.